### PR TITLE
feat: relations, schemas, and sources in the text format

### PR DESCRIPTION
**Pull request by EpsilonPrime**
* Adds support for relations, schemas, and sources.
* Does not add support for pipelines and expressions.
* Changes locations for binary to text conversion to use protobuffer message pointers.

The next PR will add expressions, fields, and literals.

### CODE REVIEW

File: src/substrait/textplan/tests/ParseResultMatchers.h
Comment (westonpace) [inline src/substrait/textplan/tests/ParseResultMatchers.h:20]: Totally a nerd snipe (feel free to ignore) but I wonder if the two `Serialize` matches could be combined to...

```
Serialized(::testing::Eq(exact_text))
Serialized(::testing::HasSubstr(contains_text))
```

**Review by westonpace — CHANGES_REQUESTED** [review]
Some initial thoughts (I don't think anything major).

File: src/substrait/textplan/Location.cpp
Comment (EpsilonPrime) [reply src/substrait/textplan/Location.cpp:48]: Fixed.

File: src/substrait/textplan/converter/PlanPrinterVisitor.cpp
Comment (EpsilonPrime) [reply src/substrait/textplan/converter/PlanPrinterVisitor.cpp:110]: Fixed.

File: src/substrait/textplan/SymbolTablePrinter.cpp
Comment (EpsilonPrime) [reply src/substrait/textplan/SymbolTablePrinter.cpp:52]: Done.

File: src/substrait/textplan/SymbolTablePrinter.cpp
Comment (EpsilonPrime) [reply src/substrait/textplan/SymbolTablePrinter.cpp:130]: Fixed.

File: src/substrait/textplan/SymbolTablePrinter.cpp
Comment (EpsilonPrime) [reply src/substrait/textplan/SymbolTablePrinter.cpp:187]: Merged.

File: src/substrait/textplan/converter/InitialPlanProtoVisitor.cpp
Comment (EpsilonPrime) [reply src/substrait/textplan/converter/InitialPlanProtoVisitor.cpp:30]: Since this is supposed to be the string form of the enum I've changed it to be rel.  I have yet to see any plans using this kind of PlanRel.

File: src/substrait/textplan/converter/InitialPlanProtoVisitor.cpp
Comment (EpsilonPrime) [reply src/substrait/textplan/converter/InitialPlanProtoVisitor.cpp:63]: Works for me.

File: src/substrait/textplan/converter/InitialPlanProtoVisitor.cpp
Comment (EpsilonPrime) [reply src/substrait/textplan/converter/InitialPlanProtoVisitor.cpp:89]: Yes, the plan is definitely there before the visitor.  Afterwards the symbol table still has references to the plan and should be kept for the short time until it is turned into its final form.  That said the final representation of the symbol table in order to be agnostic to both text and proto will need to completely contain a representation of the plan and thus it won't matter if the plan exists after visiting is completed.

File: src/substrait/textplan/converter/PlanPrinterVisitor.cpp
Comment (EpsilonPrime) [reply src/substrait/textplan/converter/PlanPrinterVisitor.cpp:94]: Done.

File: src/substrait/textplan/converter/tests/BinaryToTextPlanConversionTest.cpp
Comment (EpsilonPrime) [reply src/substrait/textplan/converter/tests/BinaryToTextPlanConversionTest.cpp:144]: The first read is the type, the second is its name (which would be read2 for the second occurrence of a read in the plan).  I fiddled a bit here making this read more like English but I'm not thrilled about the difference in order to other constructs(compare this relation to the source "source local_files local").  I suspect this is one of the areas that will be quickly changed in the 1.0.1 version of the format based on user feedback but it is not clear as to the final implementation.

File: src/substrait/textplan/tests/ParseResultMatchers.h
Comment (EpsilonPrime) [reply src/substrait/textplan/tests/ParseResultMatchers.h:20]: It doesn't work out of the box but I do suspect that such a thing is possible.  For now as I don't use SerializeContains in this PR I've removed it and I'll see what I can do for the PR that actually uses it.

File: src/substrait/textplan/converter/PlanPrinterVisitor.cpp
Comment (EpsilonPrime) [reply src/substrait/textplan/converter/PlanPrinterVisitor.cpp:87]: We should be safe here as base_schema isn't part of a oneof.  Test on ubuntu to make sure.

File: src/substrait/textplan/Any.h
Comment (EpsilonPrime) [reply src/substrait/textplan/Any.h:13]: I wasn't sure which ones I needed so I implemented them all.  It turns out I currently only need one variant (but I suspect I will need to add another one later on which I can do at that time).

File: src/substrait/textplan/Any.h
Comment (EpsilonPrime) [reply src/substrait/textplan/Any.h:21]: Agreed, I was trying to handle potential implementations that I wasn't aware of as there seems to be some variety.  I've removed this instance.

File: src/substrait/textplan/Any.h
Comment (EpsilonPrime) [reply src/substrait/textplan/Any.h:65]: The problem that this wrapper addresses is the lack of location provided in the exceptions that it does raise (and shouldn't be able to because it's not a macro).  If you get back an invalid any_cast exception without a location (and if you're under a googletest the stack trace will be swallowed unless you know where it's happening) it's really hard to debug.  With this wrapper the output contains the file location which you can click on and be right where the problem occurred.  Without it I believe the steps are:

1. enter the offending test case (probably by debugging just that test)
2. step over the parse and the matcher steps to see if the any_cast exception is there
3. examine the stack trace

Any changes to the ecosystem often yield any_cast errors (I've changed the type being stored for some representations a few times and it is easy to miss some calls).  If the module is going to be maintainable by someone besides the author I believe this wrapper is required.  I agree that it's not a desirable situation which is probably one point in favor of using status objects instead of system level exceptions.

Addendum:  An alternative would be to check the any_cast to see if it contains what is it supposed to contain instead of using a blind any_cast call (allowing the user to catch the issue before causing an exception).  The only disadvantage there is that it will end up making calls more verbose unless we use a macro (or macros) to hide the check/cast.

File: src/substrait/textplan/SymbolTablePrinter.cpp
Comment (EpsilonPrime) [reply src/substrait/textplan/SymbolTablePrinter.cpp:22]: Done here.  Other places will need to wait until after all pending PRs are finished to avoid merge conflicts.

File: src/substrait/textplan/converter/tests/BinaryToTextPlanConversionTest.cpp
Comment (EpsilonPrime) [reply src/substrait/textplan/converter/tests/BinaryToTextPlanConversionTest.cpp:149]: I've pulled in the changes from the next PR to use raw strings in the test input here.  However the test output isn't using them because I would like to use a whitespace ignoring compare for the output so it won't matter what the formatting is here in the tests.  SerializeContains was partially addressing this issue so I will investigate ignoring whitespace when I revisit that implementation later on.

File: src/substrait/textplan/SymbolTablePrinter.cpp
Comment (EpsilonPrime) [reply src/substrait/textplan/SymbolTablePrinter.cpp:189]: This is where the fun begins.  I am open to switching to use std::endl throughout.  It will affect the tests as different platforms will now be emitting different line endings.  Windows is generally okay with the lack of a carriage return but it often depends on the editor.  I believe it was pre-Unix Macintoshes that used only a CR as an end of line marker would have significant trouble here.  The other consideration is how shareable do we want the resulting text output.

File: src/substrait/textplan/SymbolTablePrinter.cpp
Comment (EpsilonPrime) [reply src/substrait/textplan/SymbolTablePrinter.cpp:84]: Switched.  This form was more inline with the types in plans I had found in the wild which apparently predates this grammar.

File: src/substrait/textplan/SymbolTablePrinter.cpp
Comment (EpsilonPrime) [reply src/substrait/textplan/SymbolTablePrinter.cpp:183]: There are two ways to approach this.  The first would be to invent some sort of escapes for the text format.  There are languages that do this.  The other approach would be to fail elsewhere (during the parsing most likely).  The latter makes this language more restrictive and the former affects the code and the grammar.  I'd like to table the handling of this issue to a future point so I've added a TODO to track this issue for now.

File: src/substrait/textplan/converter/PlanPrinterVisitor.cpp
Comment (EpsilonPrime) [reply src/substrait/textplan/converter/PlanPrinterVisitor.cpp]: Done.

File: src/substrait/textplan/Any.h
Comment (westonpace) [reply src/substrait/textplan/Any.h:23]: Got it.  So the challenge in this case is that `any_cast` is both potentially exception and very frequent in the code.  Hopefully we can avoid that case going forwards.

File: src/substrait/textplan/converter/tests/BinaryToTextPlanConversionTest.cpp
Comment (westonpace) [inline src/substrait/textplan/converter/tests/BinaryToTextPlanConversionTest.cpp:144]: Why not an `R"()"` string here?  Is that because of the line endings?

File: src/substrait/textplan/tests/CMakeLists.txt
Comment (westonpace) [inline src/substrait/textplan/tests/CMakeLists.txt:23]: Newline here to end the file.
```suggestion
        gtest_main)

```

File: src/substrait/textplan/tests/SymbolTableTest.cpp
Comment (westonpace) [inline src/substrait/textplan/tests/SymbolTableTest.cpp:143]:
```suggestion
} // namespace io::substrait::textplan

```

**Review by westonpace — APPROVED** [review]
Some minor nits and then we can merge

File: src/substrait/textplan/converter/tests/BinaryToTextPlanConversionTest.cpp
Comment (EpsilonPrime) [reply src/substrait/textplan/converter/tests/BinaryToTextPlanConversionTest.cpp:144]: I was worried about whitespace (I wasn't planning on indenting) but now that you mention it the line endings are going to be a problem.  I'll look at upgrading the matchers sooner than later so we can start using raw literals in the output.

Comment (EpsilonPrime) [reply src/substrait/textplan/converter/tests/BinaryToTextPlanConversionTest.cpp]: Done.

File: src/substrait/textplan/tests/ParseResultMatchers.h
Comment (EpsilonPrime) [reply src/substrait/textplan/tests/ParseResultMatchers.h]: Done.

**Pull request merged**
